SpockOpaleye/src/ApiBase.hs

```diff
@@ -0,0 +1,47 @@
+{-# LANGUAGE FlexibleContexts       #-}
+{-# LANGUAGE FlexibleInstances      #-}
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE MultiParamTypeClasses  #-}
+{-# LANGUAGE OverloadedStrings      #-}
+{-# LANGUAGE TemplateHaskell        #-}
+
+module ApiBase where
+
+import           Control.Lens
+import qualified Data.Profunctor.Product.Default as D
+import           Data.Time                       (UTCTime, getCurrentTime)
+import           Database.PostgreSQL.Simple
+import           DataTypes
+import           Opaleye
+import           OpaleyeDef
+import           OpaleyeTypes
+import           Prelude                         hiding (id)
+
+create_row ::(
+    HasCreatedat columnsW (Maybe (Column PGTimestamptz)),
+    HasUpdatedat columnsW (Column PGTimestamptz),
+    D.Default Constant incoming columnsW, D.Default QueryRunner returned row)
+    => Connection -> Table columnsW returned -> incoming -> IO row
+create_row conn table item = do
+  current_time <- fmap pgUTCTime getCurrentTime
+  let itemPg = (constant item) & createdat .~ (Just current_time) & updatedat .~ (current_time)
+  fmap head $ runInsertManyReturning conn table [itemPg] (\x -> x)
+
+update_row :: (
+    HasUpdatedat haskells UTCTime
+    , D.Default Constant haskells columnsW
+    , D.Default Constant item_id (Column PGInt4)
+    , HasId haskells item_id
+    , HasId columnsR (Column PGInt4)
+    )
+    => Connection -> Table columnsW columnsR -> item_id -> haskells -> IO haskells
+update_row conn table it_id item = do
```

Comment:
I see where you are coming from. If you are using an ORM that provides an Identity map and in a language with mutable values, you can mutate the fields of the object and call the update method to sync itself with the database. 

But in Haskell, owing to it's immutable nature, this pattern is not possible. So in haskell, there cannot be a unique instance that represent the data for an entity. For example, a User record cannot be made to uniquly represent a row in the user table. Because that would require that another user record that represent the same row does not exist at the same time. 

In a ORM with an Identity map, if a = user.post and user2 = a.author, then user and user2 refer to the same object. But in haskell, user and user2 can be different records, even though they hold same values, including the ID.

This means, we cannot treat records as authoritative representations of the underlying domain objects, and hence cannot use patterns that depend on this assumptions.